File: src/simulation/ActionSelector.cs
using ElevenLegends.Data.Enums;
using ElevenLegends.Data.Models;

namespace ElevenLegends.Simulation;

/// <summary>
/// Selects which action the attacking/defending team will attempt in a tick,
/// and which player will execute it.
/// </summary>
public static class ActionSelector
{
    /// <summary>
    /// Chooses the action based on the current field zone and match context.
    /// </summary>
    public static ActionType SelectAction(FieldZone ballZone, bool hasPossession, IRng rng)
    {
        if (!hasPossession)
        {
            // Defending team attempts tackle or interception
            int roll = rng.NextInt(0, 100);
            return roll < 55 ? ActionType.Tackle : ActionType.Interception;
        }

        // Attacking team actions depend on field zone
        return ballZone switch
        {
            FieldZone.DefenseLeft or FieldZone.DefenseCenter or FieldZone.DefenseRight
                => SelectDefensiveZoneAction(rng),
            FieldZone.MidfieldLeft or FieldZone.MidfieldCenter or FieldZone.MidfieldRight
                => SelectMidfieldZoneAction(rng),
            FieldZone.AttackLeft or FieldZone.AttackCenter or FieldZone.AttackRight
                => SelectAttackZoneAction(ballZone, rng),
            _ => ActionType.Pass
        };
    }

    /// <summary>
    /// Selects the best player to execute the action from the team's starting lineup.
    /// Picks the player with the highest relevant attribute (with some RNG variance).
    /// </summary>
    public static Player SelectExecutor(
        Team team, ActionType action, MatchState state, IRng rng,
        IReadOnlyList<int>? activePlayerIds = null)
    {
        var activeSet = new HashSet<int>(
            activePlayerIds is { Count: > 0 } ? activePlayerIds : team.StartingLineup);
        var candidates = team.Players
            .Where(p => activeSet.Contains(p.Id))
            .Where(p => IsEligibleForAction(p, action, state))
            .ToList();

        if (candidates.Count == 0)
        {
            candidates = team.Players.Where(p => activeSet.Contains(p.Id)).ToList();
        }

        // Weighted selection: higher attribute = higher chance of being selected
        float totalWeight = 0f;
        var weights = new float[candidates.Count];
        for (int i = 0; i < candidates.Count; i++)
        {
            int attr = SuccessCalculator.GetPrimaryAttribute(candidates[i].Attributes, action);
            float weight = attr + rng.NextFloat(0f, 20f); // some randomness in selection
            weights[i] = Math.Max(weight, 1f);
            totalWeight += weights[i];
        }

        float roll = rng.NextFloat(0f, totalWeight);
        float cumulative = 0f;
        for (int i = 0; i < candidates.Count; i++)
        {
            cumulative += weights[i];
            if (roll <= cumulative)
                return candidates[i];
        }

        return candidates[^1];
    }

    /// <summary>
    /// Advances the ball zone. On success, the ball moves forward with some probability
    /// of staying in the same zone (realistic buildup play). On failure, ball retreats.
    /// </summary>
    public static FieldZone AdvanceBallZone(FieldZone current, bool success, IRng rng)
    {
        if (success)
        {
            // Successful action doesn't always advance — sometimes it maintains position
            int advanceRoll = rng.NextInt(0, 100);

            return current switch
            {
                // From defense: 60% advance, 40% stay
                FieldZone.DefenseLeft => advanceRoll < 60
                    ? ChooseRandom(rng, FieldZone.MidfieldLeft, FieldZone.MidfieldCenter)
                    : current,
                FieldZone.DefenseCenter => advanceRoll < 60
                    ? FieldZone.MidfieldCenter
                    : current,
                FieldZone.DefenseRight => advanceRoll < 60
                    ? ChooseRandom(rng, FieldZone.MidfieldRight, FieldZone.MidfieldCenter)
                    : current,
                // From midfield: 30% advance, 70% stay (realistic buildup)
                FieldZone.MidfieldLeft => advanceRoll < 30
                    ? ChooseRandom(rng, FieldZone.AttackLeft, FieldZone.AttackCenter)
                    : current,
                FieldZone.MidfieldCenter => advanceRoll < 30
                    ? ChooseRandom(rng, FieldZone.AttackLeft, FieldZone.AttackCenter, FieldZone.AttackRight)
                    : current,
                FieldZone.MidfieldRight => advanceRoll < 30
                    ? ChooseRandom(rng, FieldZone.AttackRight, FieldZone.AttackCenter)
                    : current,
                // Already in attack zone — stay
                _ => current
            };
        }
        else
        {
            // Failed action: ball retreats
            return current switch
            {
                FieldZone.AttackLeft => ChooseRandom(rng, FieldZone.MidfieldLeft, FieldZone.MidfieldCenter),
                FieldZone.AttackCenter => FieldZone.MidfieldCenter,
                FieldZone.AttackRight => ChooseRandom(rng, FieldZone.MidfieldRight, FieldZone.MidfieldCenter),
                FieldZone.MidfieldLeft => ChooseRandom(rng, FieldZone.DefenseLeft, FieldZone.DefenseCenter),
                FieldZone.MidfieldCenter => FieldZone.DefenseCenter,
                FieldZone.MidfieldRight => ChooseRandom(rng, FieldZone.DefenseRight, FieldZone.DefenseCenter),
                // Already in defense zone — stay
                _ => current
            };
        }
    }

    private static ActionType SelectDefensiveZoneAction(IRng rng)
    {
        int roll = rng.NextInt(0, 100);
        return roll < 80 ? ActionType.Pass : ActionType.Dribble;
    }

    private static ActionType SelectMidfieldZoneAction(IRng rng)
    {
        int roll = rng.NextInt(0, 100);
        if (roll < 60) return ActionType.Pass;
        if (roll < 80) return ActionType.Dribble;
        if (roll < 95) return ActionType.Cross;
        return ActionType.Shot; // long shot — rare
    }

    private static ActionType SelectAttackZoneAction(FieldZone zone, IRng rng)
    {
        int roll = rng.NextInt(0, 100);

        // Wing zones favor crosses
        if (zone is FieldZone.AttackLeft or FieldZone.AttackRight)
        {
            if (roll < 40) return ActionType.Cross;
            if (roll < 55) return ActionType.Pass;
            if (roll < 75) return ActionType.Dribble;
            return ActionType.Shot;
        }

        // Center zone: balanced with moderate shot chance
        if (roll < 30) return ActionType.Shot;
        if (roll < 55) return ActionType.Pass;
        if (roll < 80) return ActionType.Dribble;
        return ActionType.Cross;
    }

    private static bool IsEligibleForAction(Player player, ActionType action, MatchState state)
    {
        // GK shouldn't be selected for shots (unless extreme situation)
        if (player.PrimaryPosition == Position.GK && action is ActionType.Shot or ActionType.Dribble)
            return false;

        // Defenders shouldn't shoot unless in attack zone
        if (player.PrimaryPosition is Position.CB or Position.LB or Position.RB
            && action == ActionType.Shot
            && state.BallZone is not (FieldZone.AttackLeft or FieldZone.AttackCenter or FieldZone.AttackRight))
            return false;

        return true;
    }

    private static FieldZone ChooseRandom(IRng rng, params FieldZone[] options)
    {
        int index = rng.NextInt(0, options.Length - 1);
        return options[index];
    }
}

File: engine/simulation/action-selector.ts
import { ActionType, FieldZone, Position } from "@/engine/enums";
import type { MatchState, Player, Team } from "@/engine/models";
import type { IRng } from "./rng";
import { getPrimaryAttribute } from "./success-calculator";

/**
 * Selects which action the attacking/defending team will attempt in a tick,
 * and which player will execute it.
 */

/** Chooses the action based on the current field zone and match context. */
export function selectAction(
  ballZone: FieldZone,
  hasPossession: boolean,
  rng: IRng,
): ActionType {
  if (!hasPossession) {
    // Defending team attempts tackle or interception
    const roll = rng.nextInt(0, 100);
    return roll < 55 ? ActionType.Tackle : ActionType.Interception;
  }

  // Attacking team actions depend on field zone
  switch (ballZone) {
    case FieldZone.DefenseLeft:
    case FieldZone.DefenseCenter:
    case FieldZone.DefenseRight:
      return selectDefensiveZoneAction(rng);
    case FieldZone.MidfieldLeft:
    case FieldZone.MidfieldCenter:
    case FieldZone.MidfieldRight:
      return selectMidfieldZoneAction(rng);
    case FieldZone.AttackLeft:
    case FieldZone.AttackCenter:
    case FieldZone.AttackRight:
      return selectAttackZoneAction(ballZone, rng);
    default:
      return ActionType.Pass;
  }
}

/**
 * Selects the best player to execute the action from the team's starting lineup.
 * Picks the player with the highest relevant attribute (with some RNG variance).
 */
export function selectExecutor(
  team: Team,
  action: ActionType,
  state: MatchState,
  rng: IRng,
  activePlayerIds?: readonly number[],
): Player {
  const activeSet = new Set(
    activePlayerIds != null && activePlayerIds.length > 0
      ? activePlayerIds
      : team.startingLineup,
  );
  let candidates = team.players.filter(
    (p) => activeSet.has(p.id) && isEligibleForAction(p, action, state),
  );

  if (candidates.length === 0) {
    candidates = team.players.filter((p) => activeSet.has(p.id));
  }

  // Weighted selection: higher attribute = higher chance of being selected
  let totalWeight = 0;
  const weights: number[] = new Array(candidates.length);
  for (let i = 0; i < candidates.length; i++) {
    const attr = getPrimaryAttribute(candidates[i].attributes, action);
    const weight = attr + rng.nextFloat(0, 20); // some randomness in selection
    weights[i] = Math.max(weight, 1);
    totalWeight += weights[i];
  }

  const roll = rng.nextFloat(0, totalWeight);
  let cumulative = 0;
  for (let i = 0; i < candidates.length; i++) {
    cumulative += weights[i];
    if (roll <= cumulative) {
      return candidates[i];
    }
  }

  return candidates[candidates.length - 1];
}

/**
 * Advances the ball zone. On success, the ball moves forward with some probability
 * of staying in the same zone (realistic buildup play). On failure, ball retreats.
 */
export function advanceBallZone(
  current: FieldZone,
  success: boolean,
  rng: IRng,
): FieldZone {
  if (success) {
    // Successful action doesn't always advance -- sometimes it maintains position
    const advanceRoll = rng.nextInt(0, 100);

    switch (current) {
      // From defense: 60% advance, 40% stay
      case FieldZone.DefenseLeft:
        return advanceRoll < 60
          ? chooseRandom(rng, FieldZone.MidfieldLeft, FieldZone.MidfieldCenter)
          : current;
      case FieldZone.DefenseCenter:
        return advanceRoll < 60 ? FieldZone.MidfieldCenter : current;
      case FieldZone.DefenseRight:
        return advanceRoll < 60
          ? chooseRandom(
              rng,
              FieldZone.MidfieldRight,
              FieldZone.MidfieldCenter,
            )
          : current;
      // From midfield: 30% advance, 70% stay (realistic buildup)
      case FieldZone.MidfieldLeft:
        return advanceRoll < 30
          ? chooseRandom(rng, FieldZone.AttackLeft, FieldZone.AttackCenter)
          : current;
      case FieldZone.MidfieldCenter:
        return advanceRoll < 30
          ? chooseRandom(
              rng,
              FieldZone.AttackLeft,
              FieldZone.AttackCenter,
              FieldZone.AttackRight,
            )
          : current;
      case FieldZone.MidfieldRight:
        return advanceRoll < 30
          ? chooseRandom(rng, FieldZone.AttackRight, FieldZone.AttackCenter)
          : current;
      // Already in attack zone -- stay
      default:
        return current;
    }
  } else {
    // Failed action: ball retreats
    switch (current) {
      case FieldZone.AttackLeft:
        return chooseRandom(
          rng,
          FieldZone.MidfieldLeft,
          FieldZone.MidfieldCenter,
        );
      case FieldZone.AttackCenter:
        return FieldZone.MidfieldCenter;
      case FieldZone.AttackRight:
        return chooseRandom(
          rng,
          FieldZone.MidfieldRight,
          FieldZone.MidfieldCenter,
        );
      case FieldZone.MidfieldLeft:
        return chooseRandom(
          rng,
          FieldZone.DefenseLeft,
          FieldZone.DefenseCenter,
        );
      case FieldZone.MidfieldCenter:
        return FieldZone.DefenseCenter;
      case FieldZone.MidfieldRight:
        return chooseRandom(
          rng,
          FieldZone.DefenseRight,
          FieldZone.DefenseCenter,
        );
      // Already in defense zone -- stay
      default:
        return current;
    }
  }
}

function selectDefensiveZoneAction(rng: IRng): ActionType {
  const roll = rng.nextInt(0, 100);
  return roll < 80 ? ActionType.Pass : ActionType.Dribble;
}

function selectMidfieldZoneAction(rng: IRng): ActionType {
  const roll = rng.nextInt(0, 100);
  if (roll < 60) return ActionType.Pass;
  if (roll < 80) return ActionType.Dribble;
  if (roll < 95) return ActionType.Cross;
  return ActionType.Shot; // long shot -- rare
}

function selectAttackZoneAction(zone: FieldZone, rng: IRng): ActionType {
  const roll = rng.nextInt(0, 100);

  // Wing zones favor crosses
  if (zone === FieldZone.AttackLeft || zone === FieldZone.AttackRight) {
    if (roll < 40) return ActionType.Cross;
    if (roll < 55) return ActionType.Pass;
    if (roll < 75) return ActionType.Dribble;
    return ActionType.Shot;
  }

  // Center zone: balanced with moderate shot chance
  if (roll < 30) return ActionType.Shot;
  if (roll < 55) return ActionType.Pass;
  if (roll < 80) return ActionType.Dribble;
  return ActionType.Cross;
}

function isEligibleForAction(
  player: Player,
  action: ActionType,
  state: MatchState,
): boolean {
  // GK shouldn't be selected for shots (unless extreme situation)
  if (
    player.primaryPosition === Position.GK &&
    (action === ActionType.Shot || action === ActionType.Dribble)
  ) {
    return false;
  }

  // Defenders shouldn't shoot unless in attack zone
  if (
    (player.primaryPosition === Position.CB ||
      player.primaryPosition === Position.LB ||
      player.primaryPosition === Position.RB) &&
    action === ActionType.Shot &&
    state.ballZone !== FieldZone.AttackLeft &&
    state.ballZone !== FieldZone.AttackCenter &&
    state.ballZone !== FieldZone.AttackRight
  ) {
    return false;
  }

  return true;
}

function chooseRandom(rng: IRng, ...options: FieldZone[]): FieldZone {
  const index = rng.nextInt(0, options.length - 1);
  return options[index];
}
